delocate/delocating.py

```diff
@@ -277,38 +393,54 @@ def delocate_path(tree_path, lib_path,
         is a file in the path depending on ``copied_lib_path``, and the value
         is the ``install_name`` of ``copied_lib_path`` in the depending
         library.
+
+    Raises
+    ------
+    DependencyNotFound
+        When any dependencies can not be located.
     """
     if lib_filt_func == "dylibs-only":
         lib_filt_func = _dylibs_only
+    elif isinstance(lib_filt_func, str):
+        raise TypeError('lib_filt_func string can only be "dylibs-only"')
+    if lib_filt_func is None:
+        lib_filt_func = (lambda _: True)
     if not exists(lib_path):
         os.makedirs(lib_path)
-    lib_dict = tree_libs(tree_path, lib_filt_func)
-    if copy_filt_func is not None:
-        lib_dict = dict((key, value) for key, value in lib_dict.items()
-                        if copy_filt_func(key))
-    copied = delocate_tree_libs(lib_dict, lib_path, tree_path)
-    return copy_recurse(lib_path, copy_filt_func, copied)
+
+    lib_dict = {}  # type: Dict[Text, Dict[Text, Text]]
+    for library_path in walk_directory(tree_path, lib_filt_func):
+        for depending_path, install_name in get_dependencies(library_path):
+            if copy_filt_func and not copy_filt_func(depending_path):
+                continue
+            lib_dict.setdefault(depending_path, {})
+            lib_dict[depending_path][library_path] = install_name
+
+    return delocate_tree_libs(lib_dict, lib_path, tree_path)
 
 
 def _merge_lib_dict(d1, d2):
+    # type: (Dict[Text, Dict[Text, Text]], Mapping[Text, Mapping[Text, Text]]) -> None  # noqa: E501
     """ Merges lib_dict `d2` into lib_dict `d1`
     """
     for required, requirings in d2.items():
         if required in d1:
             d1[required].update(requirings)
         else:
-            d1[required] = requirings
+            d1[required] = dict(requirings)
```

Comment:
Why `dict` here?   The other branch does not copy...

Reply:
`requirings` here is a `Mapping[Text, Text]` type and must be converted into a Dict to be compatible with assigning to `d1`.

The type hinting I added here revealed an edge-case where `d1` could be given references to values in `d2` which could mean changes made to `d1` could also affect `d2`.  I assumed that wasn't intended.

Reply:
```suggestion
            # Convert to dict to satisfy type checking.
            d1[required] = dict(requirings)
```